File: work/DSA/c++/recursion_1/allIndicesOfANumber.cpp
#include <bits/stdc++.h>
using namespace std;

void helper(int input[], int n, int x, int output[], int &count, int currentIndex)
{
    if (n == 0)
    {
        return;
    }
    else if (input[0] == x)
    {
        output[count++] = currentIndex;
    }
    helper(input + 1, n - 1, x, output, count, currentIndex + 1);
}

int allIndexes(int input[], int n, int x, int output[])
{
    int count = 0;
    helper(input, n, x, output, count, 0);
    return count;
}

int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n;
    cin >> n;

    int *input = new int[n];

    for (int i = 0; i < n; i++)
    {
        cin >> input[i];
    }

    int x;

    cin >> x;

    int *output = new int[n];

    int size = allIndexes(input, n, x, output);
    for (int i = 0; i < size; i++)
    {
        cout << output[i] << " ";
    }

    delete[] input;

    delete[] output;
}

File: work/DSA/javascript/recursion_1/allIndicesOfANumber.js
let n = 5
let input = [9, 8, 10, 8, 8]
let x = 8
let output = new Array(n)


function allIndexes(input, n, x, output) {
    let count = 0
    function helper(currentIndex) {
        if (n == 0) {
            return;
        }
        --n;
        helper(currentIndex + 1);
        if (input[currentIndex] == x) {
            output[count++] = currentIndex
        }
    }
    helper(0)
    return count
}

let answer = allIndexes(input, n, x, output)
for (let i = 0; i < answer; i++) {
    console.log(output[i])
}